### ISSUE TITLE
Detecting Any in generic types and variables

### ISSUE THREAD

**Issue by ItsDrike** [astral-sh/ruff#5871]
This is a continuation from issue #5803, which assumed the lack of support for this is a bug, which it was not, making it a feature request.

The proposal here is that currently, the `flake8-annotations` rules only detect the `Any` type when used in function arguments, or their return types directly. I think it would be beneficial to also add new linter rules (or extend the existing ANN401, however that would make it different from the original handling in `flake8-annotations`).

Here's a recap of the issue:

### `Any` in generic arguments

```python
from typing import Any

def foo(x: dict[str, Any]) -> None:  # Should be an error
    ...
    
def bar() -> list[Any]:  # Should be an error
     ...
 ```
 
I think there is value in having a linter rule capable of finding use of `Any` in these kinds of nested type definitions too. Even though they are a much more precise than just pure `Any`, there is still a lot of ambiguity being introduced there, and there's no good reason for it.

If a function can really accept any kind of values, it should use `object` in the annotation, not `Any`. As an example: `Mapping[str, object]` would be a much better option than `Mapping[str, Any]`, because the type checker would later not allow you to blindly access any variable or name from that any type, while still allowing any kinds of objects in.

### `Any` in variable annotations

```python
from typing import Any, ClassVar

x: Any = 5  # Should be an error

class Foo:
    X: ClassVar[Any] = 10  # Should be an error
    Y: Any = 50  # Should be an error
```

There is no reason to only disallow the use of `Any` in functions, but not in variables. It brings in just as much type ambiguity as in functions.

**Comment by mrcljx** [comment]
This rule sounds similar to https://typescript-eslint.io/rules/no-explicit-any/ and also assumed `ANN401` (aka `any-type`) would work this way.

> The `any` type in TypeScript is a dangerous "escape hatch" from the type system. Using `any` disables many type checking rules and is generally best used only as a last resort or when prototyping code. This rule reports on explicit uses of the `any` keyword as a type annotation.
> 
> Preferable alternatives to any include:
>
> - If the `type` is known, describing it in an `interface` or `type`
> - If the `type` is not known, using the safer `unknown` type

I would really love if it existed for Ruff, because it's so easy to _accidentally_ use `Any`: sometimes the brain thinks "any object" and instead of using `object` (Python's obscure variant to TS' `unknown`) one writes `Any` without realizing the implicatons.

I was contemplating banning import of `typing.Any` completely but:

1. managing `noqa` (for the rare case where it might actually be fine) often is frustrating,
2. once ignored there, would allow free-use throughout the file,
3. doesn't allow adding easy file-based exception in `ruff.toml` (for test suites).

It would save me lot code reviews comment writing if Ruff could **always** call out `list[Any]` and `dict[str, Any]` and don't need this to be _perfect_ (i.e. handling aliases or other things Ruff might not be able to determine because it doesn't type check).

**Comment by dlewis-orennia** [comment]
I just came across this today. As a developer coming from typescript and trying to enforce better coding standards on my team, this was an unexpected result. We assumed ANN401 was catching all `Any` that showed up only to have our AI review bot catch it.

This is even more important now with AI tools becoming so prevalent and wanting to ensure the strictest standards for when utilizing agents for assisted development.

Hoping for a solution to this. Thanks!

**Comment by jw-ciridae** [comment]
My team would also really like to see this rule added. This and a check to prevent generics in type annotations similar to mypy's `--disallow-any-generics` in ruff or ty
